### PR TITLE
Isolated SMT arrays for logically seperate storage

### PR DESCRIPTION
**Pull request by d-xo**
## Description

This pr adds an smt level optimization that generates a unique smt array for each logically seperate region in storage (i.e. each solidity level map / array will get it's own smt level array). This allows us to completely eliminate invocations of keccak from many of our smt queries, which gives a huge speedup on all storage related operations. 

The code here is still somewhat primitive, and seems to choke on more complex cases involving multiple reads / writes from nested arrays, but already gives a very satisfying speedup on storage related benchmarks. 

## Notes
- Found and fixed a minor bug in `readStorage` -- the `a` and `b` of Lit + Array / Array + Lit were copy-pasted and the guard conditions were hence swapped around.
- Array separation does not work currently. Added a NOTE to `readStorage` that indicates why. It also leads to `decompose-3-mixed` is failing. Basically, a of potentially large offset allowing full overwrite of the storage. Maybe we need to discuss about this?
- The fuzz testing `load-decompose` is currently turned off because we need to have more constraints on the generated query, or the counterexamples it gives as error are actually fine.

## Checklist

- [x] tested locally
- [ ] added automated tests
- [ ] updated the docs
- [ ] updated the changelog

### CODE REVIEW

File: src/EVM/Expr.hs [hunk 1]
```diff
@@ -644,7 +644,11 @@ readStorage w st = go (simplify w) st
       -- for our purposes. This lets us completely simplify reads from write
       -- chains involving writes to arrays at literal offsets.
       (Lit a, Add (Lit b) (Keccak _) ) | a < 256, b < maxW32 -> go slot prev
-      (Add (Lit a) (Keccak _) , Lit b) | a < 256, b < maxW32 -> go slot prev
+      (Add (Lit a) (Keccak _) , Lit b) | b < 256, a < maxW32 -> go slot prev
+
+      --- NOTE these are needed to succeed in rewriting arrays with a variable index
+      -- (Lit a, Add (Keccak _) (Var _) ) | a < 256 -> go slot prev
+      -- (Add (Keccak _) (Var _) , Lit b) | b < 256 -> go slot prev
```
Comment (d-xo) [inline src/EVM/Expr.hs:651]: do we need these?

File: src/EVM/Expr.hs [hunk 2]
```diff
@@ -756,14 +760,148 @@ writeStorage key val store = SStore key val store
...
+--   - all base stores must be one of:
+--     - concrete + empty
+--     - concrete + all keys < 256
+--     - abstract
+--   - all read / write indices must be one of:
+--     - Lit < 256
+--     - MappingSlot
+--     - ArraySlotWithOffset
+--     - ArraySlotZero
+--  - mixing within the same store & load is not permitted
```
[the remark this answers is not in the document]
Comment (msooseth) [reply src/EVM/Expr.hs:788]: Good point. I have tired :)

File: src/EVM/Expr.hs [hunk 3]
```diff
@@ -756,14 +760,148 @@ writeStorage key val store = SStore key val store
...
+data StorageType = SmallSlot | Array | Map | Mixed | UNK
+  deriving (Show, Eq)
+
+newtype CollectStorageType = CollectStorageType {storeType :: StorageType }
+  deriving (Show)
+
+initCollectStorageType :: CollectStorageType
+initCollectStorageType = CollectStorageType {storeType = UNK }
+
+decomposeCheckTypes :: Expr a -> Bool
```
[the remark this answers is not in the document]
Comment (msooseth) [reply src/EVM/Expr.hs]: Yes, but we check that too, later!

```
      if idx == idx2 then Just (SStore key2 v b) else Nothing
```

If the `idx`-es don't match, we have a Nothing. I have now tried to clarify this in the comment above.

**Review by msooseth — COMMENTED** [review]
Thank you for all the comments and pointers! I think it's better now, also it's a lot cleaner thanks to your help with the  `State StoratgeType ()` -- I learned something! I'd be curious what you think of it now :)

**Pull request merged**